File: common/dir.py
from pos import Pos
from enum import Enum

Dir = Enum('Dir', list('NESW'))

def dir_reverse(d):
    match d:
        case Dir.N: return Dir.S
        case Dir.E: return Dir.W
        case Dir.S: return Dir.N
        case Dir.W: return Dir.E

def go(p, d):
    match d:
        case Dir.N: return Pos(p.i - 1, p.j)
        case Dir.E: return Pos(p.i, p.j + 1)
        case Dir.S: return Pos(p.i + 1, p.j)
        case Dir.W: return Pos(p.i, p.j - 1)

File: day_23/src/dir.rs
use std::fmt;

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Dir {
    U,
    R,
    D,
    L,
}

impl fmt::Display for Dir {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.as_char())
    }
}

impl Dir {
    pub fn iter() -> impl Iterator<Item = Dir> {
        [Dir::U, Dir::R, Dir::D, Dir::L].into_iter()
    }

    pub fn as_char(self) -> char {
        match self {
            Dir::U => 'U',
            Dir::R => 'R',
            Dir::D => 'D',
            Dir::L => 'L',
        }
    }
}
